File: compiler/bench/cse2.fs
\ expected: 900000000
\ More complex CSE - same expensive computation multiple times

: expensive ( n -- n ) dup * dup * ;

: process ( n -- result )
  expensive   \ n^4
  dup +       \ 2*n^4
  swap drop ;

: main
  0 100000000 0 do
    3 process drop 9 +
  loop . cr ;

File: compiler/bench/cse2.c
// expected: 900000000
// More complex CSE - same expensive computation multiple times
#include <stdio.h>

static long expensive(long n) {
    long sq = n * n;
    return sq * sq;
}

static long process(long n) {
    long e = expensive(n);
    return e + e;
}

int main() {
    long sum = 0;
    for (long i = 0; i < 100000000; i++) {
        process(3);
        sum += 9;
    }
    printf("%ld\n", sum);
    return 0;
}
